### PR TITLE
feat: add support for numpy scalar types

### PR DESCRIPTION
**Pull request by techytushar**
Signed-off-by: Tushar Mittal <chiragmittal.mittal@gmail.com>

# TL;DR

* Add support for numpy scaler types
* Update type matching in the type engine
* Add tests

## Type
 - [x] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

* Add support for numpy scaler types
* Update type matching in the type engine
* Add tests

## Tracking Issue
Fixes: [https://github.com/flyteorg/flyte/issues/2753](https://github.com/flyteorg/flyte/issues/2753)

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by techytushar** [comment]
@samhita-alla Have created a new PR with tests

File: flytekit/types/schema/types.py [hunk 1]
```diff
@@ -320,7 +320,7 @@ class FlyteSchemaTransformer(TypeTransformer[FlyteSchema]):
         _np.float32: SchemaType.SchemaColumn.SchemaColumnType.FLOAT,
         _np.float64: SchemaType.SchemaColumn.SchemaColumnType.FLOAT,
         float: SchemaType.SchemaColumn.SchemaColumnType.FLOAT,
-        _np.bool: SchemaType.SchemaColumn.SchemaColumnType.BOOLEAN,  # type: ignore
+        _np.bool_: SchemaType.SchemaColumn.SchemaColumnType.BOOLEAN,  # type: ignore
```
Comment (samhita-alla) [inline flytekit/types/schema/types.py:323]: Why have you changed this? Any specific reason?

**Comment by samhita-alla** [comment]
I'm not sure if we should opt for a targeted approach or ensure all the NumPy scalars are serializable. The latter case is handled by this PR which uses `joblib` for serialization. As for the targeted approach, we can use [`Primitive`](https://github.com/flyteorg/flytekit/blob/caf612d37e3767d6d0184d6d4e86c4a01f8adeab/flytekit/models/literals.py#L48) to build the `Literal` but it might not be able to handle all the NumPy scalars and every scalar type (at least the sub-groups) need to be handled separately.

**Comment by pingsutw** [comment]
yup, It's better to convert numpy scalars to the primitive type, just like we do in [structured dataset](https://github.com/flyteorg/flytekit/blob/caf612d37e3767d6d0184d6d4e86c4a01f8adeab/flytekit/types/structured/structured_dataset.py#L309-L312)

**Comment by samhita-alla** [comment]
@techytushar, can you modify your code to return primitive types rather than serializing data directly? This might take some time, so if you aren't able to get that done before Oct. 31st, I can still add the "hacktoberfest-accepted" label and it should count towards your hacktoberfest PRs.

File: flytekit/types/schema/types.py (hunk 1 above)
Comment (techytushar) [reply flytekit/types/schema/types.py:323]: While running tests, numpy was giving warning:
```
`np.bool` is a deprecated alias for the builtin `bool`.  If you specifically wanted the numpy scalar type, use `np.bool_` here.
```
so have updated it

**Comment by techytushar** [comment]
@samhita-alla sure, I can look into it, until then please add the label

**Comment by eapolinario** [comment]
Cleaning stale PRs. Please reopen if you wan to discuss this further.
